src/util.rs

```diff
@@ -12,87 +16,289 @@ pub(crate) fn recursive_sort_plist_keys(plist: &mut plist::Dictionary) {
+    if result.len().saturating_add(suffix.len()) > MAX_LEN {
+        let mut boundary = MAX_LEN.saturating_sub(suffix.len());
         while !result.is_char_boundary(boundary) {
             boundary -= 1;
         }
         result.truncate(boundary);
     }
+
+    // Replace trailing periods and spaces by underscores unless we have a
+    // suffix (which we asserted is safe).
+    if suffix.is_empty() && result.ends_with(['.', ' ']) {
+        let mut boundary = result.len();
+        for (i, c) in result.char_indices().rev() {
+            if c != '.' && c != ' ' {
+                break;
+            }
+            boundary = i;
+        }
+        let underscores = "_".repeat(result.len() - boundary);
+        result.replace_range(boundary..result.len(), &underscores);
+    }
+
+    // Test for reserved names and parts. The relevant part is the prefix + name
+    // (or "stem") of the file, so e.g. "com1.glif" would be replaced by
+    // "_com1.glif", but "hello.com1.glif", "com10.glif" and "acom1.glif" stay
+    // as they are. For algorithmic simplicity, ignore the presence of the
+    // suffix and potentially replace more than we strictly need to.
+    //
+    // List taken from
+    // <https://docs.microsoft.com/en-gb/windows/win32/fileio/naming-a-file#naming-conventions>.
+    static SPECIAL_RESERVED: &[&str] = &[
+        "con", "prn", "aux", "nul", "com1", "com2", "com3", "com4", "com5", "com6", "com7", "com8",
+        "com9", "lpt1", "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
+    ];
+    if let Some(stem) = result.split('.').next() {
+        // At this stage, we only need to look for lowercase matches, as every
+        // uppercase letter will be followed by an underscore, automatically
+        // making the name safe.
+        if SPECIAL_RESERVED.contains(&stem) {
+            result.insert(0, '_');
+        }
+    }
+
     result.push_str(suffix);
-    result
+
+    // Test for clashes. Use a counter with 15 digits to look for a name not yet
+    // taken. The UFO specification lists a second way, should one exhaust the
+    // 15 digits, but it is unlikely to be needed in practice.
+    if existing.contains(&result.to_lowercase()) {
+        // First, cut off the suffix (plus the space needed for the number
+        // counter if necessary).
+        const NUMBER_LEN: usize = 15;
+        if result.len().saturating_sub(suffix.len()).saturating_add(NUMBER_LEN) > MAX_LEN {
+            let mut boundary = MAX_LEN.saturating_sub(suffix.len()).saturating_sub(NUMBER_LEN);
+            while !result.is_char_boundary(boundary) {
+                boundary -= 1;
+            }
+            result.truncate(boundary);
+        } else {
+            // Cutting off the suffix should land on a `char` boundary.
+            result.truncate(result.len().saturating_sub(suffix.len()));
+        }
+
+        let mut found_unique = false;
+        for counter in 1..1_000_000_000_000_000u64 {
```

Comment:
I am flabbergasted that this is the recommended strategy for determining a name. 

using some napkin math: if we assume the average length of an existing name is 8 bytes (and we need 15 bytes to store at least the numbers, for the larger numbers) then you would need something like 7000 terabytes of ram in order to just keep the *existing* names in memory.

similarly: if each check takes a microsecond (and I assume it's longer) checking all 10^15 possible names would take ~27 hours.

Reply:
I simply copied the approach from the spec, see https://unifiedfontobject.org/versions/ufo3/conventions/#common-user-name-to-file-name-algorithm. I agree that it's trying a bit hard.

I had the idea of appending a UUIDv4 and calling it a day, that'd also solve the existing paths set problem. But uh.

Reply:
yea, I see that this is the spec, and I think it's fine to follow it! I just think it's silly that this is what the spec demands.

Reply:
I think it's more of a suggestion. The spec itself notes that there is no definitive scheme.

Reply:
if it's a suggestion then let's just stop at 1000 or so, since if this situation is happing it's probably a bug and it's better to crash then to hang.

Reply:
Thinking about this more: we *do* already have the UUID crate, and I think you were on a good track when you mentioned using that. It should really simplify the code, and hopefully also help binary size?

what I would try here:

- do the 'increment' trick for a small number of integers (9? 99?)
- if that fails, try with a UUID
- if that fails, panic.


If you don't like this then we don't have to do it, but it feels much cleaner to me?

Reply:
I'll try a 100 times and then just panic. I think we can still add UUIDv4 if we ever panic in the real world :) Thanks for the suggestion!